### PR TITLE
PR1 cc3200 cpu only

### PR DESCRIPTION
**Pull request by attdona**
I dont understand why there is Makefile.include and MK60DZ10.h in the files, I have not touched these files.

### CODE REVIEW

File: sys/shell/shell.c [hunk 1]
```diff
@@ -256,6 +256,9 @@ static int readline(char *buf, size_t size)
         else {
             *line_buf_ptr++ = c;
             _putchar(c);
+#ifdef MODULE_NEWLIB
+            fflush(stdout);
+#endif
```
Comment (OlegHahm) [inline sys/shell/shell.c]: I think the last three files here (sys/shell/shell.c, drivers/Makefile.include and  doc/doxygen/riot.doxyfile
) are not related to this PR.

File: cpu/k60/include/system_MK60DZ10.h [hunk 2]
```diff
@@ -1 +0,0 @@
-system_MK60D10.h
```
Comment (OlegHahm) [inline cpu/k60/include/system_MK60DZ10.h]: How are these k60 changes related?

File: sys/shell/shell.c (hunk 1 above)
Comment (attdona) [inline sys/shell/shell.c]: sys/shell/shell.c: otherwise the shell doesn't work correctly (no echo)

doc/doxygen/riot.doxyfile: otherwise the RIOT documentation are corrupted with cc3200sdk doygen directives.

drivers/Makefile.include: I added here some directive to build the cc3200sdk unser drivers (correct me if this is the wrong way)

Comment (OlegHahm) [inline sys/shell/shell.c]: > sys/shell/shell.c: otherwise the shell doesn't work correctly

I just wonder how this related to this platform/CPU.

For the doxygen and Makefile changes, I understand the purpose, but I think they shouldn't be part of this cpu-only PR, right?

Comment (attdona) [inline sys/shell/shell.c]: about shell you are right, formally it is not related to this cpu, the important thing is that if it is a bug then it has to be pulled into.

For doxygen and Makefile they also are not part of this PR, but this PR is just a workaround ... the goal is to have all the files into the final merge and I've created the PR in the most easy way for me.

I hope this doesnt hurt too much ...

Comment (OlegHahm) [inline sys/shell/shell.c]: > about shell you are right, formally it is not related to this cpu, the important thing is that if it is a bug then it has to be pulled into.

Can you extract this into a separate PR then, please? I'm not so sure, if want to consider this as an actual "bug", because the shell is based on linebreak, so flushing every single character seems to bring little benefit.

File: Makefile.include [hunk 3]
```diff
@@ -290,6 +290,13 @@ term: $(filter flash, $(MAKECMDGOALS))
 doc:
 	make -BC $(RIOTBASE) doc
 
+btest:
```
Comment (OlegHahm) [inline Makefile.include]: What the purpose of this target?

File: core/msg.c [hunk 4]
```diff
@@ -207,7 +207,7 @@ int msg_send_int(msg_t *m, kernel_pid_t target_pid)
         return 1;
     }
     else {
-        DEBUG("msg_send_int: Receiver not waiting.\n");
+        DEBUG("msg_send_int: Receiver %d not waiting.\n", target->pid);
```
Comment (OlegHahm) [inline core/msg.c]: Unrelated to this PR and the formatter should be `PRIkernel_type` instead.

File: cpu/cc3200/Makefile.include [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+# CPU depends on the cortex-m common module, so include it:
+# include $(RIOTCPU)/cortexm_common/Makefile.include
+
+# Avoid overriding the default rule:
+all:
+
+# Rule to generate assembly listings from ELF files:
+%.lst: %.elf
+	$(OBJDUMP) $(OBJDUMPFLAGS) $< > $@
```
Comment (OlegHahm) [inline cpu/cc3200/Makefile.include]: This Makefile seems to be unnecessary long. Why not just including `cpu/Makefile.include.cortexm_common`?

File: cpu/cc3200/atomic_arch.c [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ This function executes a exclusive LDR instruction for 32 bit values.
+
+ \param [in]    ptr  Pointer to data
+ \return        value of type uint32_t at (*ptr)
+ */
+__attribute__((always_inline))   __STATIC_INLINE uint32_t __LDREXW(
+        volatile uint32_t *addr) {
+    uint32_t result;
+
+    __ASM volatile ("ldrex %0, %1" : "=r" (result) : "Q" (*addr) );
```
Comment (OlegHahm) [inline cpu/cc3200/atomic_arch.c]: Why not using the definition from `cpu/cortexm_common/include/core_cmInstr.h`?

File: cpu/cc3200/atomic_arch.c [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    __ASM volatile ("strex %0, %2, %1" : "=&r" (result), "=Q" (*addr) : "r" (value) );
+    return (result);
+}
+
+/** \brief  Remove the exclusive lock
+
+ This function removes the exclusive lock which is created by LDREX.
+
+ */
+__attribute__((always_inline)) __STATIC_INLINE void __CLREX(void) {
```
Comment (OlegHahm) [inline cpu/cc3200/atomic_arch.c]: Seems to be duplicated of `cpu/cortexm_common/atomic_arch.c`.

**Comment by OlegHahm** [comment]
I have the impression you could save a lot of code by using the implementations from `cpu/cortexm_common` instead of copying basically the same code here.

File: cpu/cc3200/include/hwtimer_cpu.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+#define HWTIMER_MAXTIMERS   TIMER_NUM_CHANNELS /**< Number of hwtimers */
+#define HWTIMER_SPEED       F_CPU   /**< The hardware timer runs at cpu clock (80MHz) */
+#define HWTIMER_MAXTICKS    0xFFFFFFFF         /**< 32-bit timer */
+    /** @} */
+
+#ifdef __cplusplus
+} /* end extern "C" */
+#endif
+
+#endif /* __HWTIMER_CPU_H */
```
Comment (OlegHahm) [inline cpu/cc3200/include/hwtimer_cpu.h]: hwtimer doesn't exist any more.

File: Makefile.include (hunk 3 above)
Comment (attdona) [inline Makefile.include]: btest is a "board unit test" that runs a python script executed by gdb python engine.

make BOARD=<board> TEST_SUITE=mysuite

will execute the python file $(CURDIR)/tests/mysuite.py

where mysuite.py is something like:

	import unittest

	int_pointer_type=gdb.lookup_type('int').pointer()

	TIMER3_TAR=0x40033048


	def mem_content(address):
		addr_val = gdb.Value(address)
		addr_ptr = addr_val.cast(int_pointer_type)
		return long(addr_ptr.dereference())

	class HwTimerTest(unittest.TestCase):

		def test_hwtimer_wait(self):
			gdb.execute("break main.c:55")
			gdb.execute("continue")
			start_time=mem_content(TIMER3_TAR)
			gdb.execute("break +1")
			gdb.execute("continue")
			end_time=mem_content(TIMER3_TAR)
			print(end_time-start_time)

			self.assertLessEqual(4000000, end_time-start_time)
			self.assertLessEqual(end_time-start_time, 4002000)

			# get SP register value
			#print(gdb.parse_and_eval("$sp"))
			#print(gdb.execute("x 0x40033048"))

	unittest.main()


This is just a beginning tentative of using python with gdb.

Comment (OlegHahm) [inline Makefile.include]: Should this be really go into the general Makefile or stay something board specific?

Comment (attdona) [inline Makefile.include]: I dont know, perhaps it is better to reject this patch and eventually fill a separate PR if this approach clearly add value to test regression tools.

File: cpu/cc3200/Makefile.include (hunk 5 above)
Comment (attdona) [inline cpu/cc3200/Makefile.include]: right! This was my first approach but I encountered problems.
I will try again ...

Comment (attdona) [inline cpu/cc3200/Makefile.include]: I confirm, I'm not able to build using cortexm_common.

The reason that discouraged me was that cc3200 lack CMSIS support:
see for example https://e2e.ti.com/support/wireless_connectivity/f/968/t/348552

I've tried again to adapt but I've to renounce, it is not for me ;)

File: sys/shell/shell.c (hunk 1 above)
Comment (attdona) [inline sys/shell/shell.c]: Ok

File: cpu/k60/include/system_MK60DZ10.h (hunk 2 above)
Comment (attdona) [inline cpu/k60/include/system_MK60DZ10.h]: On RIOT master branch system_MK60DZ10.h is a link, in my fork it is a file exactly equal to the linked file, so I think this generated the problem.
It is not clear to me if it is related to some wrong setting of mine or it is something else, very strange.

Comment (OlegHahm) [inline cpu/k60/include/system_MK60DZ10.h]: Ok, I'll provide a separate fix for that.

File: cpu/cc3200/Makefile.include (hunk 5 above)
Comment (OlegHahm) [inline cpu/cc3200/Makefile.include]: I started to adapt your base PR to the cortexm_common structure in https://github.com/OlegHahm/RIOT/tree/cc3200_launchpad. Following the example for the `ek-lm4f120xl` it shouldn't be too difficult to finish this adaption. The next thing that is currently writing a `vectors.c` and removing the `startup.c`.

Comment (attdona) [inline cpu/cc3200/Makefile.include]: Ok Oleg, nice!
Do you have a cc3200 launchpad?
Let me know when you finish the adaptation, I will test the port.

Comment (OlegHahm) [inline cpu/cc3200/Makefile.include]: I think there's a misunderstanding. I don't have the launchpad and I don't to plan to work further on it. I just wanted to give a starting point to reduce unnecessary code duplication. If there are any concrete questions, I and other people in the community are happy to help.

Comment (attdona) [inline cpu/cc3200/Makefile.include]: Ok, no problem, it's my misunderstanding

I will work on this adaptation starting with your skeleton.

Comment (OlegHahm) [inline cpu/cc3200/Makefile.include]: Cool, looking forward and always willing to help.

Comment (attdona) [inline cpu/cc3200/Makefile.include]: Ok, I've adapted to cortexm_common. I made some tests and the only thing doesn't work is the reboot function.

With NVIC_SystemReset() the board does not reset, so cortexm_common/reboot_arch implementation has to be overwritten.

I've tried to redefine reboot_arch.c inside cpu/cc3200 but the build system seems not to work with this approach.

How are the guidelines for this type of issues?

Comment (OlegHahm) [inline cpu/cc3200/Makefile.include]: Do you know why this is the case?

Comment (attdona) [inline cpu/cc3200/Makefile.include]: Not precisely ...
cc3200 is a complex soc with 3 MCU inside, so I suppose the reset machinery could be not standard (my ipothesis).

What works is the API PRCMMCUReset from sdk/driverlib:

    MCU Reset (Software Reset)
    MCU subsystem can be reset to its default state using the following API function call.
    void PRCMMCUReset(tBoolean bIncludeSubsystem)
    Description: This function performs a software reset of a MCU and associated peripherals.    The core
    resumes execution in the ROM boot loader which will re-load the user application from sFlash.
    Parameter: bIncludeSubsystem – If true, MCU along its associated peripherals are reset,

Comment (attdona) [inline cpu/cc3200/Makefile.include]: I dont know if there is a more riotic way, but something like that could work?

cpu/cc3200.h

    #define SYSTEM_RESET PRCMMCUReset(1)

reboot_arch.c

	#ifndef SYSTEM_RESET
	#define SYSTEM_RESET  NVIC_SystemReset()
	#endif

	int reboot_arch(int mode)
	{
	    printf("Going into reboot, mode %i\n", mode);
	    /* wait a while to make sure the printf is finished */
	    volatile int n = 100000;
	    while(n--);
	    SYSTEM_RESET;

	    return -1;
	}

**Comment by kYc0o** [comment]
Feature freeze -> postponed. Can you rebase and re-order the commits? They're complicated to review so if you can rename it using the coding conventions would be great!

**Comment by attdona** [comment]
Rebased. See #5733
